### PR TITLE
borgbackup: 1.2.8 -> 1.4.0

### PR DESCRIPTION
**Pull request by e1mo**
## Description of changes

Various fixes, new `BORG_EXIT_CODES=modern` and `borg version`, documentation improvements.

Changelog:
<https://github.com/borgbackup/borg/blob/1.4.0/docs/changes.rst>

> If you currently use borg 1.2.5+: no upgrade steps needed
> (if you already did them when upgrading to that version, otherwise see below).

Upgrade notes: https://github.com/borgbackup/borg/blob/1.4.0/docs/changes.rst#upgrade-notes

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/backup/borgbackup/default.nix [hunk 1]
```diff
@@ -3,6 +3,7 @@
 , acl
 , e2fsprogs
 , fetchFromGitHub
+, fetchPypi
```
Comment (bjornfor) [inline pkgs/tools/backup/borgbackup/default.nix:6]: Unused addition.

**Review by nadir-ishiguro — APPROVED** [review]

**Comment by nadir-ishiguro** [comment]
Result of `nixpkgs-review pr 324323` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>borgbackup</li>
    <li>borgbackup.dist</li>
    <li>borgbackup.doc</li>
    <li>borgbackup.man</li>
    <li>borgmatic</li>
    <li>borgmatic.dist</li>
    <li>emborg (python311Packages.emborg)</li>
    <li>emborg.dist (python311Packages.emborg.dist)</li>
    <li>pika-backup</li>
    <li>python312Packages.emborg</li>
    <li>python312Packages.emborg.dist</li>
    <li>vorta</li>
    <li>vorta.dist</li>
  </ul>
</details>

File: pkgs/tools/backup/borgbackup/default.nix [hunk 2]
```diff
@@ -82,7 +82,6 @@ python.pkgs.buildPythonApplication rec {
   nativeCheckInputs = with python.pkgs; [
     e2fsprogs
     py
-    python-dateutil
```
Comment (bjornfor) [inline pkgs/tools/backup/borgbackup/default.nix:85]: How did you learn about this input being unneeded?

Comment (e1mo) [reply pkgs/tools/backup/borgbackup/default.nix:85]: The [Changelog for 1.4.0a1](https://github.com/borgbackup/borg/blob/1.4.0/docs/changes.rst#version-140a1-2024-01-01) states:

> tests: remove python-dateutil dependency

Comment (bjornfor) [reply pkgs/tools/backup/borgbackup/default.nix:85]: Thanks.

**Review by bjornfor — APPROVED** [review]
LGTM. (Didn't test.)

**Comment by mweinelt** [comment]
> Is `BORG_EXIT_CODES=modern` helpful for the NixOS module?

https://github.com/borgbackup/borg/blob/1.4.0/docs/internals/frontends.rst#message-ids
https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#SuccessExitStatus=

SuccessExitStatus=100 (FileChangedWarning)

**Comment by e1mo** [comment]
> > Is `BORG_EXIT_CODES=modern` helpful for the NixOS module?
> 
> [borgbackup/borg@`1.4.0`/docs/internals/frontends.rst#message-ids](https://github.com/borgbackup/borg/blob/1.4.0/docs/internals/frontends.rst?rgh-link-date=2024-07-05T11%3A39%3A19Z#message-ids) [freedesktop.org/software/systemd/man/latest/systemd.service.html](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#SuccessExitStatus=)
> 
> SuccessExitStatus=100 (FileChangedWarning)

https://github.com/NixOS/nixpkgs/blob/9f4128e00b0ae8ec65918efeba59db998750ead6/nixos/modules/services/backup/borgbackup.nix#L36-L45

I'd rather use the `100` in `borgWrapper` and treat it as everything as expected (just like the rc 1 for generic warnings).

**Comment by bjornfor** [comment]
Since `BORG_EXIT_CODES=modern` is opt-in, I suggest we merge this PR and have a follow-up PR for the NixOS side of things.

**Review by fpletz — APPROVED** [review]

**Comment by bjornfor** [comment]
borg 1.4.x is the current stable version (https://www.borgbackup.org/releases/); should we backport it to release-24.05, after a few weeks perhaps?

**Pull request merged**